### ISSUE TITLE
[FEATURE] increase 110 pods per node limit

### ISSUE THREAD

**Issue by fanlix** [harvester/harvester#2707]
## problem
k8s has default limit:  110 pods per node.
My harvester control node with 20 vm already used 100 pods now.

####  pod numbers by namespaces:
* cattle  13 pods
* kube-system 19 pods
* longhorn 29 pods
* harvester-system 15 pods
* user vm only 24 pods 

#### harvester k8 config shows 110 limit
```
  allocatable:
    cpu: "40"
    memory: 395731228Ki
    pods: "110"
  capacity:
    cpu: "40"
    memory: 395731228Ki
    pods: "110"
```

## Describe the solution you'd like
* need an easy way to increase the limit number 110 -> 200
* or make 200 as the default harvester limit

#### for normal k8s cluster
* kubelet args: `--max-pods 200`
* kubelet config file: `/var/lib/rancher/rke2/agent/kubelet.kubeconfig`
* rancher kubelet config extra_args:  `max-pods: 250`

## Describe alternatives you've considered
any chance to reduce harvester/rancher/longhorn pods?

**Comment by tlehman** [comment]
We can make 200 the limit. I'll make a PR now

**Comment by tlehman** [comment]
The ETA for the PR: Monday Dec 5

**Comment by yasker** [comment]
We need a better understanding of how many pods we should support per node. It normally going to depend on how big the node is. Have we decided to just set it to 200 or allow users or customize it or we can automatically decide it based on some machine specs (potentially with a user override)?

**Comment by tlehman** [comment]
Good question, the [best practices doc](https://kubernetes.io/docs/setup/best-practices/cluster-large/) says "No more than 110 pods per node", but Amazon EKS has a file that defines node maxes for each EC2 type that could be useful here: https://github.com/awslabs/amazon-eks-ami/blob/master/files/eni-max-pods.txt

They have max-pods as high as 737 on a z1d.metal.

**Comment by tlehman** [comment]
If I sample a few of those EC2 types, I can generate a heuristic based on CPU/RAM/Disk that will produce a pod limit based on those resources. That would approximately reproduce the EKS limits.

**Comment by yasker** [comment]
Taking a look at https://github.com/awslabs/amazon-eks-ami/blob/master/files/eni-max-pods.txt . It seems the number has some relationship with the spec of the node.

**Comment by yasker** [comment]
> If I sample a few of those EC2 types, I can generate a heuristic based on CPU/RAM/Disk that will produce a pod limit based on those resources. That would approximately reproduce the EKS limits.

Sounds like a good idea.

**Comment by tlehman** [comment]
I used SQLite to join the ec2 types with the max pods limits in that eks file. Now I just need to feed that into a linear model that we can use to choose the max pods limit. https://gist.github.com/tlehman/34e3e2028f7fea69ae6f34faf0e625b3

**Comment by tlehman** [comment]
Using just number of vCPUs (threads), there are almost three discrete levels 
![cpu](https://user-images.githubusercontent.com/261552/206010635-38b95aba-bd14-4545-a748-a6c1bd0c25e9.png)


The histogram of the maxpods validates that: 
![hist](https://user-images.githubusercontent.com/261552/206010254-4826c0a2-246d-4342-90dd-241e99fb7c2b.png)

So I think we can make three limits, 

<100 vCPU      maxpods = 110 (default)
>100 vCPU, <=200   maxpods = 220
>=200  maxpods = 440?

It's node-specific, and it could be automatic, instead of set by the user.
Does this sound reasonable?

**Comment by tlehman** [comment]
Better idea: do a linear regression on (cpu, mem). When I included mem, the mem coefficient was slightly negative, and about 40x smaller than cpu. So I simplified it to just cpu, the result is about $y = 2.6 x + 186$ where $x$ is the cpu count, and $y$ is the maxpods for that node. Then we just take the ceiling $\lceil y \rceil$ of $y$.

![image](https://user-images.githubusercontent.com/261552/206031489-df48d6ad-ec2c-4944-b4b9-ffca8bb6935d.png)

**Comment by tlehman** [comment]
After some internal discussion, we decided to make the maxpods configurable by the user. They will be set on the host using a label, the interface will look like this:

![Screenshot from 2022-12-06 14-38-49](https://user-images.githubusercontent.com/261552/206039312-ef1f540c-c497-4368-b2c7-f12e1c094649.png)

That value will get passed into the kubelet on that host, setting the capacity to the value given by the user, or the default 110, if the label is not present.

**Comment by tlehman** [comment]
Proposed backend design:

Set the label for maxPods on the Host UI (like the screenshot), then modify the node validator to set the maxpods param in Update

https://github.com/tlehman/harvester/commit/9e5de911a4c1b4d3c22896b42f973b5023556f85#diff-0c6595ff8184b074abdb6a42aad763f355c2c3e457019c44a7213cdbd6cdeab6

**Comment by Martin-Weiss** [comment]
What will be the network design for this >110 POD limit? 
Cluster CIDR in RKE2?
And what rules do you plan to define for network ranges per node for PODs?

-> will this enhancement allow to modify the Cluster CIDR and address range per node after initial deployment?

**Comment by tlehman** [comment]
@Martin-Weiss Great question, I think following [GCE's maxpods/CIDR range table](https://cloud.google.com/kubernetes-engine/docs/how-to/flexible-pod-cidr#:~:text=The%20following%20table%20lists%20the%20size%20of%20the%20CIDR%20block%20and%20the%20corresponding%20number%20of%20available%20IP%20addresses%20that%20Kubernetes%20assigns%20to%20nodes%20based%20on%20the%20maximum%20Pods%20per%20node%3A) is the best option here:

**Comment by Martin-Weiss** [comment]
> @Martin-Weiss Great question, I think following [GCE's maxpods/CIDR range table](https://cloud.google.com/kubernetes-engine/docs/how-to/flexible-pod-cidr#:~:text=The%20following%20table%20lists%20the%20size%20of%20the%20CIDR%20block%20and%20the%20corresponding%20number%20of%20available%20IP%20addresses%20that%20Kubernetes%20assigns%20to%20nodes%20based%20on%20the%20maximum%20Pods%20per%20node%3A) is the best option here:

Yes - the challenge will be how to match this cluster wide CIDR vs. node wide subnet.. and find a way to "not overlap" with existing IP address ranges at customers where some pods might have "outgoing" communication to.. and to deal with different ranges per node might be difficult, too.

**Comment by tlehman** [comment]
My first thought about this would be to sum up the total number of pods, and then set the cluster CIDR range to include more addresses than that sum. But that's only if we want a flat topology.

This feels like it needs a proper HEP. I will write one up and then invite you to the review meeting

**Comment by ibrokethecloud** [comment]
For now we should bump up the max-pods to 200 via kubelet-args. This can also be incorporated via a patch in the upgrade path.

With the default canal settings currently in place, running more than 254 pods per node will require exposing the cluster and pod cidr ranges, and subnetLen. This will then need to be translated into rke2-canal-config, configmap.

Or we could just make the default pod and cluster cidr range larger and allow a /23 subnet per node, which could be the new default, allowing about 510 max-pods per node.

For making the max-pod dynamic, we would need additional reconcile logic in the harvester controllers, to watch node labels, to reconcile the same into the harvester-server or harvester-agent config via changes to `/oem/99-config.yaml`. 

Flannel config guide for reference: https://github.com/flannel-io/flannel/blob/master/Documentation/configuration.md

**Comment by guangbochen** [comment]
Moving the issue to the v1.2 release as this is a new feature enhancement instead of a quick config change; this will need a proper design and investigations with HEP, thanks.
Also prefer to set the maximum pod limitations to 200 for now, as it will introduce fewer configuration changes like CIDR range and upgrade capability support, we may need a global setting to be configured for all nodes but can be overwritten individually via per-node config.

**Comment by Martin-Weiss** [comment]
Is it valid to run 200 pods on a single RKE2 node with the default CIDR setup we have in harvester? Somehow I thought that the 110 pod limit is based on the default range per node based on /24.. I have in mind that we need two IP per pod but that might be wrong..

And when adding a feature for "more" - how to we want to change the CIDRs during upgrades as changing this for existing deployments has not been supported for a long time... thought it is just supported during initial cluster creation..

**Comment by jrlittlefoot64** [comment]
Hello, I have read this https://github.com/orgs/cloud-hypervisor/projects/6/views/4 ([Assess integration with KubeVirt#4201](https://github.com/cloud-hypervisor/cloud-hypervisor/issues/4201), I think it will have performance effects to increase to 200 Pods

**Comment by ibrokethecloud** [comment]
To test the change:

## New Install:
* install latest master iso.
* once cluster is up the new pod capacity can be checked as follows:
```
kubectl get nodes -o yaml | yq '.items[].status.capacity.pods'
200
200
```
* expect to find 200 for each node in the cluster.

## Upgrade path:
* upgrade from v1.1.2 to latest master iso.
* once upgrade is complete, the pod capacity can be checked as follows:
```
kubectl get nodes -o yaml | yq '.items[].status.capacity.pods'
200
200
```
* expect to find 200 for each node in the cluster.

**Comment by TachunLin** [comment]
Verified **fixed** on master-70b3fbf4-head (05/22). Close this issue.

### Result
#### [PASS] Test Case1: New install (default setting)
* On the fresh install Harvester with latest master, we can find each node of cluster have `200` pods capacity
    ```
    node1:~ # kubectl get nodes -o yaml | yq '.items[].status.capacity.pods'
    
    200
    200
    200
    ```

#### [PASS] Test Case2:  Upgrade Path (default setting) 
* After we successfully upgrade from v1.1.2 to v1.2.0 (master-70b3fbf4)
![image](https://github.com/harvester/harvester/assets/29251855/dae26245-9235-4407-a029-c25d1a53582a)

* Check each node of cluster can have `200` pods capacity
  ```
  node2:~ # kubectl get nodes -o yaml | yq '.items[].status.capacity.pods'
  200
  200
  200
  ```

### Test Information
* Test Environment: 3 nodes harvester on bare machine
* Harvester version: master-70b3fbf4-head (05/22)

### Verify Steps
https://github.com/harvester/harvester/issues/2707#issuecomment-1556966960
